File: lesson_1/code/task_2_4_malthusian_difference.m
% MODELING PARAMETERS
    P0 = 2.50;                  % Intial popoulation in milions
    r = 0.027;                  % Population Growth rate 2.7%
    start_year = 1950;          % For plotting
    t_start = 0;  t_end = 70;   % Start and end years
    t_diff = 1;                 % Step (1/12 is one month, 1/365 is one day) 

% EXPLICIT SOLUTION
    t_expl= t_start:t_diff:t_end;
    P_expl= P0*exp(r*t_expl);

% NUMERICAL SOLUTION   
   t_num = t_start;  P_num = P0; 
    for i = 1:t_end/t_diff    
           P_num = [P_num, P_num(i) + r*P_num(i)*t_diff;];   
           t_num = [t_num, t_num(i) + t_diff];
    end

% PLOTTING
tiledlayout (1, 2)
nexttile       % Tile with explicit and numerical solution
    hold on
    plot(t_expl + start_year, P_expl , Color = "red")
    plot(t_num + start_year, P_num , Color = " blue")
    title("Explicit and numerical solution")
    ylim([0,20]); xlim([1950,2020])
    hold off
nexttile       % Tile with the difference
    plot(t_expl + start_year, P_expl - P_num, Color =" black")
    title ("The difference")
    ylim([0,0.5]); xlim([1950,2020])

File: lesson_1/code/task_2_4_malthusian_difference.py
import numpy as np
import matplotlib.pyplot as plt

# MODELING PARAMETERS
P0, r = 2.50,  0.027
start_year = 1950
t_start, t_end, t_diff = 0, 70, 1

# EXPLICIT SOLUTION
t_expl = np.arange(t_start, t_end + t_diff, t_diff)
P_expl = P0 * np.exp(r * t_expl)

# NUMERICAL SOLUTION
t_num = [t_start]
P_num = [P0]
for i in range(int(t_end/t_diff)):
    P_num.append(P_num[i] + r * P_num[i] * t_diff)
    t_num.append(t_num[i] + t_diff)
P_num = np.array(P_num)
t_num = np.array(t_num)

# PLOTTING
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 5))

# Plot 1: Explicit vs Numerical
ax1.plot(t_expl + start_year, P_expl, color="red", label="Explicit")
ax1.plot(t_num + start_year, P_num, color="blue", label="Numerical")
ax1.set_title("Explicit and numerical solution")
ax1.set_ylim(0, 20)
ax1.set_xlim(1950, 2020)
ax1.legend()

# Plot 2: Difference
ax2.plot(t_expl + start_year, P_expl - P_num, color="black")
ax2.set_title("The difference")
ax2.set_ylim(0, 0.5)
ax2.set_xlim(1950, 2020)

plt.tight_layout()
plt.show()